### ISSUE TITLE
64 bit - static_assert error on build

### ISSUE THREAD

**Issue by strubelz** [mill1000/raspdif#17]
When I try to build I get the following error running on a rpi3b+:
```
echo "#define GIT_VERSION \"Commit: dadf5e6\"" > include/git_version.h-new; \
cmp -s include/git_version.h include/git_version.h-new || cp include/git_version.h-new include/git_version.h; \
rm include/git_version.h-new
clang -I include -I include/bcm283x -I /opt/vc/include -MMD  -Wall -Wno-missing-braces -c source/mailbox.c -o build/source/mailbox.c.o
source/mailbox.c:113:10: warning: cast to 'void *' from smaller integer type 'uint32_t' (aka 'unsigned int') [-Wint-to-void-pointer-cast]
  return (void*)response->tag.bus_address;
         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
1 warning generated.
clang -I include -I include/bcm283x -I /opt/vc/include -MMD  -Wall -Wno-missing-braces -c source/log.c -o build/source/log.c.o
clang -I include -I include/bcm283x -I /opt/vc/include -MMD  -Wall -Wno-missing-braces -c source/spdif.c -o build/source/spdif.c.o
clang -I include -I include/bcm283x -I /opt/vc/include -MMD  -Wall -Wno-missing-braces -c source/memory.c -o build/source/memory.c.o
source/memory.c:105:53: warning: cast to smaller integer type 'uint32_t' (aka 'unsigned int') from 'void *' [-Wvoid-pointer-to-int-cast]
  return (void*)((uint32_t)entry.pfn * pageSize) + ((uint32_t) virtual % pageSize);
                                                    ^~~~~~~~~~~~~~~~~~
1 warning generated.
clang -I include -I include/bcm283x -I /opt/vc/include -MMD  -Wall -Wno-missing-braces -c source/main.c -o build/source/main.c.o
In file included from source/main.c:11:
In file included from include/bcm283x/bcm283x.h:24:
include/bcm283x/bcm283x_dma.h:221:1: error: static_assert failed due to requirement 'sizeof(struct dma_control_block_t) == 8 * sizeof(unsigned int)' "dma_control_block_t must be 32 bytes."
static_assert(sizeof(dma_control_block_t) == 8 * sizeof(uint32_t), "dma_control_block_t must be 32 bytes.");
^             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/include/assert.h:143:24: note: expanded from macro 'static_assert'
# define static_assert _Static_assert
                       ^
In file included from source/main.c:11:
In file included from include/bcm283x/bcm283x.h:24:
include/bcm283x/bcm283x_dma.h:223:1: error: static_assert failed due to requirement 'sizeof(struct bcm283x_dma_channel_t) == 9 * sizeof(unsigned int)' "bcm283x_dma_channel_t must be 36 bytes."
static_assert(sizeof(bcm283x_dma_channel_t) == 9 * sizeof(uint32_t), "bcm283x_dma_channel_t must be 36 bytes.");
^             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/include/assert.h:143:24: note: expanded from macro 'static_assert'
# define static_assert _Static_assert
                       ^
2 errors generated.
make: *** [Makefile:31: build/source/main.c.o] Error 1

```

**Comment by mill1000** [comment]
Interesting. I'm guessing based on the other warnings you are running a 64-bit OS, and that's throwing off the size of the pointers in the DMA structures.

Can you provide the outputs of the following commands?

`clang --version`, `lsb_release -a`, `uname -a`

**Comment by as-cmi** [comment]
I'm running on Docker, same problem:

`root@e766abdc2321:~# clang --version
Ubuntu clang version 14.0.0-1ubuntu1
Target: aarch64-unknown-linux-gnu
Thread model: posix
InstalledDir: /usr/bin`

`root@e766abdc2321:~# lsb_release -a
LSB Version:    core-11.1.0ubuntu4-noarch:security-11.1.0ubuntu4-noarch
Distributor ID: Ubuntu
Description:    Ubuntu 22.04.2 LTS
Release:        22.04
Codename:       jammy`

`root@e766abdc2321:~# uname -a
Linux e766abdc2321 6.1.19-v8+ #1637 SMP PREEMPT Tue Mar 14 11:11:47 GMT 2023 aarch64 aarch64 aarch64 GNU/Linux`

**Comment by mill1000** [comment]
| Linux e766abdc2321 6.1.19-v8+ #1637 SMP PREEMPT Tue Mar 14 11:11:47 GMT 2023 **aarch64** aarch64 aarch64 GNU/Linux

Yep, that's a 64 bit architecture. For now, that's not supported. I'd have to think about the intricacies of supporting it. e.g. the DMA peripheral still expects 32 bit addresses, but the OS is accessing the same memory via 64 bit addresses.

As for Docker. I'd be surprised if it's possible, but glad to see you try. raspdif has to access some low-level hardware which I don't know how it will interact with Docker's isolation.
Some thoughts to consider:
* You'll need to bind mount `/dev/mem` and `/dev/vcio` into the container.
* raspdif uses some Raspberry Pi specific functions from `bcm_host`. These might not be available in an Ubuntu base image.

**Comment by as-cmi** [comment]
I know i'll likely encounter problems either with docker isolation and maybe performance issues.
I was planning to compile using docker and maybe run the executable on the host.
Thanks for your work!
Looking forward for a 64 bit version :D

**Comment by mill1000** [comment]
I have a branch for [64 bit support here](https://github.com/mill1000/raspdif/tree/feature/64bit_support). It's working on my RPi3.

I'm more or less blindly casting things to 32 bit as needed. Quite possible it might fail on HW with more than 4 GB of mem.

**Comment by mill1000** [comment]
Closed via #22
